File: 01-easy/0087-query-board/0087-query-board.py
#!/usr/bin/env python
# -*- coding: utf-8; py-indent-offset:4 -*-
###############################################################################
#
# Copyright (C) 2016 Daniel Rodriguez
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.
#
###############################################################################
from __future__ import (absolute_import, division, print_function,)
#                        unicode_literals)

import sys

# Small Py2/3 compatibility layer
if sys.version_info.major == 2:
    range = xrange
else:  # >= 3
    pass


if __name__ == '__main__':
    test_cases = open(sys.argv[1], 'r')

    BDIM = 256
    BSIZE = BDIM * BDIM
    board = [0] * BSIZE

    for line in test_cases:
        ls = line.rstrip('\n')

        if ls.startswith('Set'):
            cmd, coord, val = ls.split()
            coord, val = int(coord), int(val)

            if cmd[3:] == 'Col':
                r = range(coord, BSIZE, BDIM)
            else:
                r = range(coord * BDIM, coord * BDIM + BDIM)

            for x in r:
                board[x] = val
        else:
            cmd, coord = ls.split()
            coord = int(coord)
            if cmd[5:] == 'Col':
                r = range(coord, BSIZE, BDIM)
            else:
                r = range(coord * BDIM, coord * BDIM + BDIM)

            result = sum(board[x] for x in r)
            print(result)

    test_cases.close()

File: 01-easy/0087-query-board/0087-query-board.cpp
/*
  Copyright (C) 2015, 2016 Daniel Rodriguez

  This program is free software: you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation, either version 3 of the License, or
  (at your option) any later version.

  This program is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

// Headers for test case input/output
#include <fstream>
#include <iostream>

#include <cstring>

///////////////////////////////////////////////////////////////////////////////
// Main
///////////////////////////////////////////////////////////////////////////////
int
main(int argc, char *argv[]) {
    std::ifstream stream(argv[1]);

    const int BLEN = 256;
    const int BSIZE = BLEN * BLEN;
    int board[BSIZE] = {};  // init to default entire array (0)

    char cmd[10];
    int coord, val;

    while(stream >> cmd) {
        if(::strncmp(cmd, "Set", 3) == 0) {
            stream >> coord >> val;
            if(::strncmp(&cmd[3], "Row", 3) == 0) {
                for(auto i=coord * BLEN; i < coord * BLEN + BLEN; i++)
                    board[i] = val;
            } else {
                for(auto i=coord; i < BSIZE; i += BLEN)
                    board[i] = val;
            }
        } else {
            stream >> coord;
            val = 0;
            if(::strncmp(&cmd[5], "Row", 3) == 0) {
                for(auto i=coord * BLEN; i < coord * BLEN + BLEN; i++)
                    val += board[i];
            } else {
                for(auto i=coord; i < BSIZE; i += BLEN)
                    val += board[i];
            }
            std::cout << val << std::endl;
        }
    }
    return 0;
}
